### ISSUE TITLE
Visualization in Stage1 and Stage2

### ISSUE THREAD

**Issue by zhumanli** [dimitrismallis/KeypointsToLandmarks#1]
Hi,

Thank you for open source the codes!

I am wondering why there is almost no difference in the visualization between stage 1 and stage 2. I'm using the pre-trained model with 10 keypoints for visualization.

**Comment by dimitrismallis** [comment]
Hi, 

this is rather expected behaviour. 

Given that CelebA comprises mostly frontal poses, when K is set to a smaller value like 10, our method tends to consistently detect all 10 landmarks in each image since they are rarely occluded. If you experiment with larger values of K (for example 30), you will observe more missing landmarks in stage 1 compared to stage 2 due to factors like occlusion and self-occlusion etc.

**Comment by zhumanli** [comment]
> 

Thank you very much for your answer!

I'm also wondering whether the contrastive learning process is robust, for example, the cluster labels of keypoints might change in different clustering iterations, will this cause any confusion in the network? I see you update the pseudo-ground truth every 5000 iterations, does this try to make the contrastive learning more consistent?

**Comment by dimitrismallis** [comment]
Hi,

Indeed the clustering labels of keypoints will change for different clustering rounds. Kmeans does not guarantee the stability of clustering assignments. This will not affect landmark representation learning through the contrastive loss. During training of Stage 1, we only use clustering assignments to form positive and negative pairs. Two landmarks belonging to the same cluster always form an accurate positive pair even though the cluster's label can be different for subsequent clustering rounds.

**Comment by zhumanli** [comment]
Thank you so much for your explanation! It makes sense.
